lib/Logger.jsx

```diff
@@ -21,57 +21,49 @@ export default class Logger {
-    * @param {Number} recentMessages A number of recent messages to append as context
-    * @param {Object|String} parameters The parameters to send along with the message
     */
-    logToServer(message, recentMessages, parameters = {}) {
-        // Optionally append recent log lines as context
-        let msg = message;
-        if (recentMessages > 0) {
-            msg += ' | Context:  ';
-            msg += JSON.stringify(this.get(recentMessages));
-        }
-
-        // Output the message to the console too.
-        if (this.isDebug) {
-            this.client(`${msg} - ${JSON.stringify(parameters)}`);
+    logToServer() {
+        // We do not want to call the server with an empty list or if all the lines has onlyFlushWithOthers=false
+        if (!_.any(this.logLines, l => !l.onlyFlushWithOthers)) {
+            return;
         }
 
         // We don't care about log setting web cookies so let's define it as false
-        const params = {parameters, message, api_setCookie: false};
-        this.serverLoggingCallback(params);
+        const promise = this.serverLoggingCallback({api_setCookie: false, logPacket: JSON.stringify(this.logLines)});
+        this.logLines = [];
+        if (!promise) {
+            return;
+        }
+        promise.then((response) => {
+            if (response.requestID) {
+                this.info('Previous log requestID', false, {requestID: response.requestID}, true);
+            }
+        });
     }
 
     /**
      * Add a message to the list
      * @param {String} message
      * @param {Object|String} parameters The parameters associated with the message
+     * @param {Boolean} forceFlushToServer Should we force flushing all logs to server?
+     * @param {Boolean} onlyFlushWithOthers A request will never be sent to the server if all loglines have this set to true
      */
-    add(message, parameters) {
+    add(message, parameters, forceFlushToServer, onlyFlushWithOthers = false) {
         const length = this.logLines.push({
             message,
             parameters,
+            onlyFlushWithOthers,
             timestamp: (new Date())
         });
 
-        // If we're over the limit, remove one line.
-        if (length > this.TEMP_LOG_LINES_TO_KEEP) {
-            this.logLines.shift();
+        if (this.isDebug) {
+            this.client(`${message} - ${JSON.stringify(parameters)}`);
         }
-    }
 
-    /**
-     * Get the last messageCount lines
-     * @param {Number} messageCount Number of messages to get (optional,
-     *              defaults to 1)
-     * @return {array} an array of messages (oldest first)
-     */
-    get(messageCount = 1) {
-        // Don't get more than we have
-        const count = Math.min(this.TEMP_LOG_LINES_TO_KEEP, messageCount);
-        return this.logLines.slice(this.logLines.length - count);
+        // If we're over the limit, flush the logs
+        if (length > MAX_LOG_LINES_BEFORE_FLUSH || forceFlushToServer) {
```

Comment:
Whoa TIL I learn that `.push()` returns the length of the array?! 
🤯   never ever saw this used ever before.